### PR TITLE
feat(developer): Debugger platform option 🐞

### PR DESCRIPTION
**Pull request by mcdurdin**
Relates to #5013.

Adds a debug status page to allow the user to select an alternative platform so they can debug rules relating to those platforms which are constrained by the `platform()` statement in the keyboard source.

Note that while touch platforms are listed here, there is not yet any method of activating touch key events (e.g. `[T_foo]`). This functionality is planned for a future update.

# User Testing

Add a keyboard with `platform()` statements such as the following:

```
platform('windows hardware desktop native') + [K_1] > 'windows hardware desktop native '
platform('windows') + [K_4] > 'windows '
platform('mac') + [K_4] > 'mac '
platform('macos') + [K_5] > 'macos '
platform('macosx') + [K_6] > 'macosx '
platform('linux') + [K_4] > 'linux '
platform('touch') + [K_4] > 'touch '
```

Start the debugger.

* TEST_WINDOWS: Verify that the baseline Windows platform is selected

  1. In the Options tab, select Windows platform, type <kbd>1</kbd>
  2. The output should be `windows hardware desktop native `

* TEST_MAC: Verify that various mac aliases are matched

  1. In the Options tab, select macOS platform
  2. Type <kbd>4</kbd>. 'mac ' should be output
  3. Type <kbd>5</kbd>. 'macos ' should be output
  4. Type <kbd>6</kbd>. 'macosx ' should be output

* TEST_OTHERS: Verify that other platform selections work

  1. Add platform rules to match each operating system, following the example above
  2. For each platform, verify that the rule matches as expected

* TEST_UI: Verify that the way that platform combo boxes are enabled and disabled as you select different options makes sense

  1.  if OS is 'macOS', 'windows', or 'linux' then 'touch' should not be available.
  2.  if OS is 'iOS' then 'desktop' and 'hardware' options should not be available (we don't support hardware keyboards or desktop environments on iOS)
  3. if OS is 'Android' then 'desktop' option should not be available.
  4. Note that the Browser list shows all browsers, even on platforms where those browser may not be available. This is not a failure condition.

### CODE REVIEW

**Comment by MakaraSok** [comment]
- **TEST_WINDOWS: FAILED** System Error. Code: 126.
This is the same error as noted in issue https://github.com/keymanapp/keyman/issues/5648. 
   image

   Crash reports:
   ```
   Details for error report for Keyman app:
  Application Identifier: tike-15.0.104-alpha-test-5640
  Error Identifier: 497C085E4ADA46572CCA811E4F285761
   ```

- **TEST_MAC: BLOCKED** 
- **TEST_OTHERS: BLOCKED** 
- **TEST_UI: BLOCKED**

File: windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    begin
+      cbPlatformUI.ItemIndex := PLATFORM_UI_HARDWARE;
+      cbPlatformUI.Enabled := False;
+      if cbPlatformFormFactor.ItemIndex = PLATFORM_FF_DESKTOP then
+        cbPlatformFormFactor.ItemIndex := PLATFORM_FF_TABLET;
+      cbPlatformFormFactor.Enabled := True;
+    end;
+    PLATFORM_OS_ANDROID:
+    begin
+      cbPlatformUI.Enabled := True;
```
Comment (darcywong00) [inline windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas:130]: Why is this `True` when PLATFORM_OS_IOS is `False`?

I also would have expected PLATFORM_OS_WINDOWS to have True too.

**Comment by mcdurdin** [comment]
@keymanapp-test-bot retest

Based on the fixes in #5649, the new build should include kmnkbp0-0.dll so the debugger should start.

**Comment by MakaraSok** [comment]
The same issue is experienced:
```
Details for error report for Keyman app:
  Application Identifier: tike-15.0.104-alpha-test-5640
  Error Identifier: 1AC6E134BE3B4A288EAC9AA68A3EB549
```

* **TEST_WINDOWS: BLOCKED** see #5648
Consequently, the rest of the test cases cannot be carried out either.
* **TEST_MAC: BLOCKED**
* **TEST_OTHERS: BLOCKED**
* **TEST_UI: BLOCKED**

File: windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    PLATFORM_OS_WINDOWS, PLATFORM_OS_MAC, PLATFORM_OS_LINUX:
+    begin
+      cbPlatformUI.ItemIndex := PLATFORM_UI_HARDWARE;
+      cbPlatformUI.Enabled := False;
+      cbPlatformFormFactor.ItemIndex := PLATFORM_FF_DESKTOP;
+      cbPlatformFormFactor.Enabled := False;
+    end;
+    PLATFORM_OS_IOS:
+    begin
+      cbPlatformUI.ItemIndex := PLATFORM_UI_HARDWARE;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas:122]:
```suggestion
      cbPlatformUI.ItemIndex := PLATFORM_UI_TOUCH;
```

File: windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas (hunk 1 above)
Comment (mcdurdin) [reply windows/src/developer/TIKE/debug/UfrmDebugStatus_Options.pas:130]: Only Android devices let us switch between Hardware and Touch modes. I note that the iOS one should have been `TOUCH` not `HARDWARE`...

**Comment by mcdurdin** [comment]
@keymanapp-test-bot retest

Based on the fixes in #5649, the new build should include kmnkbp0-0.dll so the debugger should start. Apologies, I didn't merge the changes into this branch, last time...

**Comment by MakaraSok** [comment]
- **TEST_WINDOWS: BLOCKED** "7" doesn't output the expected string, but itself. "1" does though. 
   Mark this as BLOCKED for now as there might be something wrong with the instructions.
- **TEST_MAC: PASSED** 456 outputs the expected "mac macos macosx".
- **TEST_OTHERS: PASSED** All works as expected.
- **TEST_UI: FAILED** Everything in the UI seems OK, iOS.
   1. the UI doesn't show the dropdown menu of "Hardware" and "Touch" as those in the Android.
   2. Form Factor for iOS includes "Desktop". There might be no Desktop running iOS. If so, it doesn't make sense to have it win the list.

**Comment by mcdurdin** [comment]
@keymanapp-test-bot retest TEST_UI

**Comment by MakaraSok** [comment]
* **TEST_WINDOWS: PASSED** `windows hardware desktop native` does get output when key <kbd>1</kbd> was pressed.
* **TEST_UI: FAILED** 

> 1. if OS is 'macOS', 'windows', or 'linux' then 'touch' should not be available.

This seems OK. 'Touch' is indeed not available on these OSes.

> 2. if OS is 'iOS' then 'desktop' and 'hardware' options should not be available (we don't support hardware keyboards or desktop environments on iOS)

When 'iOS' was selected, the UI dropdown menu is grey out with "Hardware" visible. To me, this should have been "Touch" as it is really what is set by default.


> 3. if OS is 'Android' then 'desktop' option should not be available.

'desktop' option actually is available in the dropdown menu of UI options.



> 4. Note that the Browser list shows all browsers, even on platforms where those browser may not be available. This is not a failure condition.

Windows:
   

macOS:
   

Linux:
   

For Android/iOS, see above screenshots.

_Please note that the Browser menu is grey out without having anything visible on any OS while it should be shown a list of available browsers._

**Comment by mcdurdin** [comment]
> When 'iOS' was selected, the UI dropdown menu is grey out with "Hardware" visible. To me, this should have been "Touch" as it is really what is set by default.

Fixed. Good pick!

> > 3. if OS is 'Android' then 'desktop' option should not be available.
> 
> 'desktop' option actually is available in the dropdown menu of UI options.

Your screenshot shows 'hardware' not 'desktop'?

> Please note that the Browser menu is grey out without having anything visible on any OS while it should be shown a list of available browsers.

You need to select 'Web' instead of 'Native' under the 'Application' option, in order to select a web browser.

@keymanapp-test-bot retest TEST_UI

**Comment by MakaraSok** [comment]
> Your screenshot shows 'hardware' not 'desktop'?

Wrong screenshot, however the current build looks OK.

* **TEST_UI: PASSED** Everything looks OK.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.114-alpha](https://keyman.com/downloads/releases/alpha/15.0.114)

**Pull request merged**